### PR TITLE
Support date function: split out builtin catalog by item type

### LINKED ISSUE #14812: sql: `COLLECTION_SYSTEM_GID_MAPPING` loses distinction between items and functions
### What version of Materialize are you using?

main

### How did you install Materialize?

Built from source

### What is the issue?

`COLLECTION_SYSTEM_GID_MAPPING` relies on a slightly-too-simplistic version of name resolution when handling `GlobalId`s; namely, there is an "invisible" namespace distinction between functions and all other items in a schema. Elsewhere in the catalog, we handle this appropriately by resolving items in either `schema.items` or `schema.functions`. However, only the schema and the item name are stored in `COLLECTION_SYSTEM_GID_MAPPING`, which means you cannot differentiate between looking up either an item or a function.

This issue prevents us from supporting functions with the same name as a function because we cannot have a `GlobalId` for both the function and the object; instead, the more recently inserted object will be the one that claims the `GlobalId`, in essence dropping the other.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by chaas**
Add new parameter to the `COLLECTION_SYSTEM_GID_MAPPING` key in the catalog for the item type, so that objects of different types can have the same name.
This allows us to have a function with the same name as a type, which is required to support the `date` function cast. Postgres supports function-like syntax casts, which Looker uses for `date`. #14803

The rest of the adapter correctly handles functions separately from types.
Resolves #14812.

This PR  also adds the `date` function to `src/sql/src/func.rs` that calls our existing cast functions.

This PR also removes the "hack" solution temporarily implemented in #14824 .

### Tips for reviewer
Let me know if there's a better way to get the `CatalogItemType` from the `Builtin` struct. The `catalog_item_type ` function feels more hardcoded than should be needed.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [inline src/adapter/src/catalog/storage.rs:1472]: @sploiselle Mentioned in a discussion that if we added this we'd also need to add some code that did some namespace enforcement. Is that still the case, or does the SQL layer do a good enough job of that?

Comment (sploiselle) [reply src/adapter/src/catalog/storage.rs:1472]: Yeah, what I was suggesting is essentially something that would have caught the error that this fixes. This will come up again if we ever try to create a builitn object with the same name as a type.

The advantage of this is that it won't take someone in the future an hour to find the problem.

This is also only necessary for builtin types specifically because they _don't_ go through the SQL layer upon creation, so there's nothing preventing us from bungling the naming rules.


```diff
diff --git a/src/adapter/src/catalog.rs b/src/adapter/src/catalog.rs
index 422f2f35a..3eaa94dd2 100644
--- a/src/adapter/src/catalog.rs
+++ b/src/adapter/src/catalog.rs
@@ -472,13 +472,19 @@ impl CatalogState {
             .flatten()
     }
 
+    /// Associates a name, `GlobalId`, and entry. Returns the previous
+    /// `GlobalId` associated with that name, if any.
+    ///
+    /// The returned value is useful only in the case of builtin types, which
+    /// do not otherwise go through name resolution to ensure the desired name
+    /// is not occupied.
     fn insert_item(
         &mut self,
         id: GlobalId,
         oid: u32,
         name: QualifiedObjectName,
         item: CatalogItem,
-    ) {
+    ) -> Option<GlobalId> {
         if !id.is_system() && !item.is_placeholder() {
             info!(
                 "create {} {} ({})",
@@ -527,13 +533,16 @@ impl CatalogState {
             &entry.name().qualifiers.schema_spec,
             conn_id,
         );
-        if let CatalogItem::Func(_) = entry.item() {
-            schema.functions.insert(entry.name.item.clone(), entry.id);
+
+        let prev_id = if let CatalogItem::Func(_) = entry.item() {
+            schema.functions.insert(entry.name.item.clone(), entry.id)
         } else {
-            schema.items.insert(entry.name.item.clone(), entry.id);
-        }
+            schema.items.insert(entry.name.item.clone(), entry.id)
+        };
 
-        self.entry_by_id.insert(entry.id, entry.clone());
+        self.entry_by_id.insert(entry.id, entry);
+
+        prev_id
     }
 
     #[tracing::instrument(level = "trace", skip(self))]
@@ -1892,12 +1901,12 @@ impl<S: Append> Catalog<S> {
                 },
                 item: builtin.name().into(),
             };
-            match builtin {
+            let prev_id = match builtin {
                 Builtin::Log(log) => {
                     let oid = catalog.allocate_oid().await?;
                     catalog
                         .state
-                        .insert_item(id, oid, name.clone(), CatalogItem::Log(log));
+                        .insert_item(id, oid, name.clone(), CatalogItem::Log(log))
                 }
 
                 Builtin::Table(table) => {
@@ -1913,7 +1922,7 @@ impl<S: Append> Catalog<S> {
                             conn_id: None,
                             depends_on: vec![],
                         }),
-                    );
+                    )
                 }
 
                 Builtin::View(view) => {
@@ -1933,7 +1942,7 @@ impl<S: Append> Catalog<S> {
                             )
                         });
                     let oid = catalog.allocate_oid().await?;
-                    catalog.state.insert_item(id, oid, name, item);
+                    catalog.state.insert_item(id, oid, name, item)
                 }
 
                 Builtin::Type(_) => unreachable!("loaded separately"),
@@ -1945,7 +1954,7 @@ impl<S: Append> Catalog<S> {
                         oid,
                         name.clone(),
                         CatalogItem::Func(Func { inner: func.inner }),
-                    );
+                    )
                 }
 
                 Builtin::StorageCollection(coll) => {
@@ -1955,9 +1964,15 @@ impl<S: Append> Catalog<S> {
                         oid,
                         name.clone(),
                         CatalogItem::StorageCollection(coll),
-                    );
+                    )
                 }
-            }
+            };
+
+            assert!(
+                prev_id.is_none(),
+                "builiin name collision on {:?}",
+                builtin.name()
+            );
         }
         let new_system_id_mappings = new_builtins
             .iter()
@@ -2226,7 +2241,7 @@ impl<S: Append> Catalog<S> {
         // Insert into catalog
         for typ in builtin_types {
             let element_id = name_to_id_map[typ.name];
-            self.state.insert_item(
+            let prev_id = self.state.insert_item(
                 element_id,
                 typ.oid,
                 QualifiedObjectName {
@@ -2242,6 +2257,12 @@ impl<S: Append> Catalog<S> {
                     depends_on: vec![],
                 }),
             );
+
+            assert!(
+                prev_id.is_none(),
+                "builiin name collision on {:?}",
+                typ.name
+            );
         }
 
         let new_system_id_mappings = new_builtins
```

File: src/adapter/src/catalog/storage.rs
Comment (sploiselle) [inline src/adapter/src/catalog/storage.rs:535]: My mental model if this is more like `(String, CatalogItemType, String)`, where `CatalogItemType` is reifying the otherwise-invisible namespace delineation in Postgres' catalog, i.e. how functions and types can share names. This isn't a big deal, but if you see any value in it, I think reordering these will help folks in the future build a clearer intuition for why this exists.

File: src/adapter/src/catalog.rs
Comment (sploiselle) [inline src/adapter/src/catalog.rs:1648]: Can we get a comment here about the way that this impacts namespacing?

File: test/sqllogictest/cast.slt
Comment (sploiselle) [inline test/sqllogictest/cast.slt:123]: Ah, this error is insufficiently specified.

We need something akin to this diff:

```diff
diff --git a/src/pgrepr/src/oid.rs b/src/pgrepr/src/oid.rs
index e492c5fb9..b717c4f18 100644
--- a/src/pgrepr/src/oid.rs
+++ b/src/pgrepr/src/oid.rs
@@ -203,3 +203,4 @@ pub const FUNC_MZTIMESTAMP_GTE_MZTIMESTAMP_OID: u32 = 16_559;
 pub const FUNC_MZ_NOW_OID: u32 = 16_560;
 pub const FUNC_MAX_MZTIMESTAMP_OID: u32 = 16_561;
 pub const FUNC_MIN_MZTIMESTAMP_OID: u32 = 16_562;
+pub const FUNC_DATE_FROM_TEXT: u32 = 16_563;
diff --git a/src/sql/src/func.rs b/src/sql/src/func.rs
index 13ce08e42..59b5bac7c 100644
--- a/src/sql/src/func.rs
+++ b/src/sql/src/func.rs
@@ -1874,6 +1874,7 @@ pub static PG_CATALOG_BUILTINS: Lazy<HashMap<&'static str, Func>> = Lazy::new(||
             params!(Int32) => UnaryFunc::Chr(func::Chr), 1621;
         },
         "date" => Scalar {
+            params!(String) => UnaryFunc::CastStringToDate(func::CastStringToDate), oid::FUNC_DATE_FROM_TEXT;
             params!(Timestamp) => UnaryFunc::CastTimestampToDate(func::CastTimestampToDate), 2029;
             params!(TimestampTz) => UnaryFunc::CastTimestampTzToDate(func::CastTimestampTzToDate), 1178;
         },
```

In this case we need to allocate a new oid because PG doesn't handle this case as a function; they must have some special case that sniffs out typename-looking functions with a parameter of an unknown type (which means it looks like a string) and instead does a parse.

This then lets us return the correct error here, which is:

```
invalid input syntax for type date
```

(without this diff it tries to cast the parameter to a ts)

Think this is important to do because this implementation is likely to be the template everyone follows for adding these functions in the future and better to propagate this now.

**Comment by chaas** [comment]
Wow thank you @sploiselle for the thorough review! Really appreciate it.

File: src/adapter/src/catalog/storage.rs
Comment (chaas) [reply src/adapter/src/catalog/storage.rs:1472]: Ah yeah that's a good idea. Best to have that just in case. Thanks for typing it out

File: test/sqllogictest/cast.slt
Comment (chaas) [reply test/sqllogictest/cast.slt:123]: Ah that's a great suggestion

File: src/adapter/src/catalog/storage.rs
Comment (chaas) [reply src/adapter/src/catalog/storage.rs:535]: Yeah definitely agreed, I'll rearrange, good point.

File: src/adapter/src/catalog.rs
Comment (benesch) [inline src/adapter/src/catalog.rs:480]: Are you ever meant to do anything with the returned global ID besides assert that it is not none? ISTM that assertion should be pushed into this function, so that each caller doesn't need to perform the assertion independently!

File: src/adapter/src/catalog.rs
Comment (benesch) [inline src/adapter/src/catalog.rs:1657]: Wait, this explanation doesn't seem quite right to me? You can't have a table and an index with the same name, for example, even though they're different types, right?

File: src/adapter/src/catalog.rs
Comment (sploiselle) [reply src/adapter/src/catalog.rs:480]: Oh--good call! I had assumed because this check wasn't in place that this function was meant to handle updates or be idempotent, but from the looks of things that is not the case.

File: src/adapter/src/catalog.rs
Comment (sploiselle) [reply src/adapter/src/catalog.rs:1657]: Yes, that is what the check we added to `insert_item` does.

**Comment by benesch** [comment]
No need to wait for my approval here!

**Comment by chaas** [comment]
@sploiselle Made all the suggested changes - do you want to take another pass? Hoping to get this in by eod for the release.

**Review by sploiselle — APPROVED** [review]
very cool; ty Cara

**Pull request merged**
